test/test_ops.py

```diff
@@ -463,6 +439,115 @@ def _fn(*sample_args, **sample_kwargs):
+        #   Expected behavior: resize with a warning.
+        wrong_shape = list(expected.shape)
+        wrong_shape[-1] = wrong_shape[-1] + 1
+        out = make_tensor(wrong_shape,
+                          dtype=expected.dtype,
+                          device=expected.device)
+        with self.assertWarnsRegex(UserWarning, "An output with one or more elements"):
+            op_out(out=out)
+        self.assertEqual(expected, out)
+
+        # Case 3: out= with the correct dtype and device, but an empty
+        #   tensor.
+        #   Expected behavior: resize without warning.
+        # TODO: this currently asserts the op throws NO warnings,
+        #   which may be too strict (if the op is deprecated, for example)
+        out = make_tensor((0,),
+                          dtype=expected.dtype,
+                          device=expected.device)
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            op_out(out=out)
+        self.assertEqual(len(w), 0)
+        self.assertEqual(expected, out)
+
+        # Case 4: out= with correct shape and dtype, but wrong device.
+        #   NOTE: this case only tested if a CUDA device is available.
+        #   Expected behavior: error.
+        if torch.device(device).type == 'cuda':
+            # Case 4A: CUDA input with CPU out.
+            out = make_tensor(expected.shape,
+                              dtype=expected.dtype,
+                              device='cpu')
+            with self.assertRaises(RuntimeError):
+                op_out(out=out)
+
+            # Case 4B: CPU input with CUDA out.
+            #   TODO: not implemented.
+
+        # Case 5: out= with correct shape and device, but a dtype
+        #   that the output can be "safely" cast to (cdouble).
+        #   Expected behavior: same computation is performed.
+        # TODO: enable this. Too many ops require that the out
+        #     tensor have the same dtype as the input tensor.
+
+        # out = make_tensor(expected.shape,
+        #                   dtype=torch.cdouble,
+        #                   device=expected.device)
+        # op_out(out=out)
+        # self.assertEqual(expected, out, exact_dtype=False)
+
+        # Case 6: out= with correct shape and device, but a dtype
+        #   that output cannot be "safely" cast to (long).
+        #   Expected behavior: error.
+        # NOTE: this case is filtered by dtype since some ops produce
+        #   bool tensors, for example, which can be safely cast to any
+        #   dtype.
+        _dtypes = floating_and_complex_types_and(torch.float16, torch.bfloat16)
+        if expected.dtype in _dtypes:
+            out = make_tensor(expected.shape,
+                              dtype=torch.long,
+                              device=expected.device)
+            with self.assertRaises(RuntimeError):
+                op_out(out=out)
+
+        # Case 7: out= where out= is given the input tensor.
```

Comment:
Remove this. We should tell people not to do this, not enforce it in code. Because enforcing it in code is cumbersome and we don't have an efficient mechanism to detect that the out tensor contains a portion of the input tensor, for example.